engine/common/insight_log/resource_insight_logs.py

```diff
@@ -0,0 +1,132 @@
+import enum
+import json
+import logging
+import re
+from abc import ABC, abstractmethod
+
+from django.apps import apps
+
+insight_logger = logging.getLogger("insight_logger")
+logger = logging.getLogger(__name__)
+
+
+class EntityEvent(enum.Enum):
+    CREATED = "created"
+    UPDATED = "updated"
+    DELETED = "deleted"
+
+
+class InsightLoggable(ABC):
+    @property
+    @abstractmethod
+    def public_primary_key(self):
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_verbal(self) -> str:
+        """
+        insight_logs_verbal returns resource name for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_type_verbal(self) -> str:
+        """
+        insight_logs_type_verbal resource type for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_serialized(self) -> dict:
+        """
+        insight_logs_serialized returns resource, serialized for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_metadata(self) -> dict:
+        """
+        insight_logs_metadata returns resource metadata which should be present in the insight_log line
+        """
+        pass
+
+
+def resource_insight_log(instance: InsightLoggable, author, event: EntityEvent, prev_state=None, new_state=None):
+    try:
+        organization = author.organization
+        DynamicSetting = apps.get_model("base", "DynamicSetting")
+        org_id_to_enable_insight_logs, _ = DynamicSetting.objects.get_or_create(
+            name="org_id_to_enable_insight_logs",
+            defaults={"json_value": []},
+        )
+        log_all = "all" in org_id_to_enable_insight_logs.json_value
+        insight_logs_enabled = organization.id in org_id_to_enable_insight_logs.json_value
+        if insight_logs_enabled or log_all:
+            tenant_id = organization.stack_id
+            author_id = author.public_primary_key
+            author = json.dumps(author.username)
+            entity_type = instance.insight_logs_type_verbal
+            try:
+                entity_id = instance.public_primary_key
+            except AttributeError:
+                entity_id = instance.id  # Fallback for entities which have no public_primary_key
+            entity_name = json.dumps(instance.insight_logs_verbal)
+            metadata = instance.insight_logs_metadata
+            log_line = f"tenant_id={tenant_id} author_id={author_id} author={author} action_type=resource action={event.value} resource_type={entity_type} resource_id={entity_id} resource_name={entity_name}"  # noqa
```

Comment:
May be move everything above into the property that returns dict, and dump on the next line?

Reply:
I'm not sure what you mean here